### PR TITLE
APP 1416 Write tests

### PR DESCRIPTION
**Pull request by Michael-A-Heuer**
## Description

This PR adds more tests to the `Multisig` class. It also adds and fixes tests in other classes, e.g, where 

```ts
expect(await contract.call(arg)).to.emit(contract, 'MyEvent').withArgs(...);
```
was accidentally written instead of
```ts
await expect(contract.call(arg)).to.emit(contract, 'MyEvent').withArgs(...);
```
which resulted in tests silently passing without checking anything (see the [`ethereum-waffle` docs](https://ethereum-waffle.readthedocs.io/en/latest/matchers.html#emitting-events)).
`ethereum-waffle`s `withArgs` matcher seems to have an issue with arrays of structs so we have to do it in a more complicated way. We should consider using `hardhat-chai-matchers` instead.

Lastly, tests were added to check for the configuration edge case of `supportThreshold=0` and `minParticipation=0` in `TokenVoting` and `AddresslistVoting` which was added to [check for a bug that was found](https://discord.com/channels/672466989217873929/1059439408270544896/1059439410694869054) but ultimately originated from the subgraph and was fixed in https://github.com/aragon/core/pull/167.

Task: [APP-1416](https://aragonassociation.atlassian.net/browse/APP-1416)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] Any dependent changes have been merged and published in downstream modules.
- [x] I ran all tests with success and extended them if necessary.
- [x] I have updated the ``CHANGELOG.md`` file in the root folder.
- [x] I have tested my code on the test network.

[APP-1416]: https://aragonassociation.atlassian.net/browse/APP-1416?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/contracts/test/voting/addresslist-voting.ts [hunk 1]
```diff
@@ -1200,5 +1220,77 @@ describe('AddresslistVoting', function () {
         expect(await voting.canExecute(id)).to.equal(true);
       });
     });
+
+    describe('An edge case with >0% support and >=0% participation required in early execution mode', async () => {
```
Comment (novaknole) [inline packages/contracts/test/voting/addresslist-voting.ts:1224]: I don't get the description well ! in the beforeEach, we got `supportThreshold` and `minParticipation` = 0.

File: packages/contracts/test/voting/token-voting.ts [hunk 2]
```diff
@@ -1117,5 +1137,88 @@ describe('TokenVoting', function () {
         expect(await voting.canExecute(id)).to.equal(true);
       });
     });
+
+    describe('An edge case with >0% support and >=0% participation required in early execution mode', async () => {
```
Comment (novaknole) [inline packages/contracts/test/voting/token-voting.ts:1141]: same here as the comment in addresslist.

File: packages/contracts/test/voting/multisig.ts [hunk 3]
```diff
@@ -90,119 +95,138 @@ describe('Multisig', function () {
...
+      ).to.be.revertedWith(
+        customError(
+          'MinApprovalsOutOfBounds',
+          addresslistLength,
+          multisigSettings.minApprovals
+        )
+      );
+    });
+
+    it('should not allow to set minApprovals would be set to 0', async () => {
```
Comment (novaknole) [inline packages/contracts/test/voting/multisig.ts:162]: grammar error.

File: packages/contracts/test/voting/multisig.ts [hunk 4]
```diff
@@ -90,119 +95,138 @@ describe('Multisig', function () {
...
+
+    it('should emit MultisigSettingsUpdated when updateMutlsigSettings gets called', async () => {
+      await expect(multisig.updateMultisigSettings(multisigSettings))
+        .to.emit(multisig, MULTISIG_EVENTS.MULTISIG_SETTINGS_UPDATED)
+        .withArgs(multisigSettings.onlyListed, multisigSettings.minApprovals);
+    });
   });
 
-  describe('Addresslisting members: ', async () => {
+  describe('Addresslisting members:', async () => {
```
Comment (novaknole) [inline packages/contracts/test/voting/multisig.ts:176]: maybe `Addresslist members`

File: packages/contracts/test/voting/multisig.ts [hunk 5]
```diff
@@ -213,118 +237,193 @@ describe('Multisig', function () {
...
+        expect(proposal.tally.addresslistLength).to.equal(1);
+        expect(proposal.actions.length).to.equal(0);
+
+        expect(await multisig.canApprove(id, signers[0].address)).to.equal(
+          true
+        );
+        expect(await multisig.canApprove(id, signers[1].address)).to.equal(
+          false
+        );
+        expect(await multisig.canApprove(1, signers[0].address)).to.equal(
```
Comment (novaknole) [inline packages/contracts/test/voting/multisig.ts:387]: I dON't ge what this tests. `1` would not be even created. Pointless. The test name suggests that it tests something else

File: packages/contracts/test/voting/multisig.ts [hunk 6]
```diff
@@ -338,140 +437,189 @@ describe('Multisig', function () {
...
-      expect(proposal.parameters.minApprovals).to.equal(minApprovals);
-      expect(await multisig.approvals(id)).to.be.eq(minApprovals);
+      it('returns `false` if the approver has already approved', async () => {
+        await multisig.connect(signers[0]).approve(id, false);
+        expect(await multisig.canApprove(id, signers[0].address)).to.be.false;
+      });
 
-      expect(await multisig.canExecute(id)).to.equal(true);
-      await expect(multisig.execute(id)).to.not.be.reverted;
+      it('returns `false` if the approver is listed', async () => {
```
Comment (novaknole) [inline packages/contracts/test/voting/multisig.ts:462]: wrong test name ? it returns true.

File: packages/contracts/test/voting/multisig.ts [hunk 7]
```diff
@@ -338,140 +437,189 @@ describe('Multisig', function () {
...
-    it('should emit MultisigSettingsUpdated when updateMutlsigSettings gets called', async () => {
-      await multisig.initialize(dao.address, addresslist(5), {
-        onlyListed,
-        minApprovals,
+      it('emits the `ProposalExecuted` and `Executed` events', async () => {
+        await multisig.connect(signers[0]).approve(id, false);
+        await multisig.connect(signers[1]).approve(id, false);
+        await multisig.connect(signers[2]).approve(id, false);
+
+        await expect(await multisig.connect(signers[3]).execute(id))
```
Comment (novaknole) [inline packages/contracts/test/voting/multisig.ts:608]: do we need `await` 2 times ?

File: packages/contracts/test/voting/multisig.ts [hunk 8]
```diff
@@ -338,140 +437,189 @@ describe('Multisig', function () {
...
+          .to.emit(dao, DAO_EVENTS.EXECUTED)
+          .to.emit(multisig, PROPOSAL_EVENTS.PROPOSAL_EXECUTED)
+          .to.not.emit(multisig, MULTISIG_EVENTS.APPROVED);
+      });
+
+      it('emits the `Approved`, `ProposalExecuted`, and `Executed` events if execute is called inside the `approve` method', async () => {
+        await multisig.connect(signers[0]).approve(id, false);
+        await multisig.connect(signers[1]).approve(id, false);
+
+        await expect(await multisig.connect(signers[2]).approve(id, true))
```
Comment (novaknole) [inline packages/contracts/test/voting/multisig.ts:618]: do we need await 2 times ?

**Review by novaknole — COMMENTED** [review]
Very Nice !!!

File: packages/contracts/test/voting/admin.ts [hunk 9]
```diff
@@ -184,7 +175,7 @@ describe('Admin plugin', function () {
       const expectedDummyResults = ['0x'];
 
       await expect(await plugin.executeProposal(dummyMetadata, dummyActions))
```
Comment (mathewmeconry) [inline packages/contracts/test/voting/admin.ts:177]: Is `await plugin....` still necessary?

Comment (mathewmeconry) [inline packages/contracts/test/voting/admin.ts:177]: Is `await plugin....` still necessary?

**Review by mathewmeconry — APPROVED** [review]

File: packages/contracts/test/voting/addresslist-voting.ts (hunk 1 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/addresslist-voting.ts]: Fixed in f16d3e7becb9668bbd24b3ed065cdbee9af2b98f

File: packages/contracts/test/voting/token-voting.ts (hunk 2 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/token-voting.ts]: Fixed in f16d3e7becb9668bbd24b3ed065cdbee9af2b98f

File: packages/contracts/test/voting/multisig.ts (hunk 3 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/multisig.ts]: Fixed in 57174b98c910188a41e81ac5457f041910a75a0b

File: packages/contracts/test/voting/multisig.ts (hunk 4 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/multisig.ts]: Improved in 0dbbd2b6ffc041ae71ed2dd7075a318897f2f9f6

File: packages/contracts/test/voting/multisig.ts (hunk 6 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/multisig.ts]: Indeed, fixed in commit 560908ca218e50cbe90f8769e05f8c17e33d6096

File: packages/contracts/test/voting/multisig.ts (hunk 5 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/multisig.ts]: Fixed in 0f663c4521f126ff0eede25de03766ce88a69293

File: packages/contracts/test/voting/multisig.ts (hunk 8 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/multisig.ts]: Fixed in 31773140e64edfb0d2ba37bb16d6defd856d0577

File: packages/contracts/test/voting/admin.ts (hunk 9 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/admin.ts]: Fixed in 31773140e64edfb0d2ba37bb16d6defd856d0577

File: packages/contracts/test/voting/multisig.ts (hunk 7 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/multisig.ts]: Fixed in 31773140e64edfb0d2ba37bb16d6defd856d0577

**Review by novaknole — APPROVED** [review]

**Review by mathewmeconry — APPROVED** [review]

File: packages/contracts/test/voting/addresslist-voting.ts [hunk 10]
```diff
@@ -1200,5 +1220,77 @@ describe('AddresslistVoting', function () {
...
+        expect(await voting.participation(id)).to.be.gte(
+          votingSettings.minParticipation
+        );
+        // worst case support can be calculated without throwing an error even if nobody has voted
+        expect(await voting.worstCaseSupport(id)).to.be.eq(
+          votingSettings.supportThreshold
+        );
+        expect(await voting.canExecute(id)).to.equal(false);
+
+        // does not execute normally
```
[the remark this answers is not in the document]
Comment (brickpop) [reply packages/contracts/test/voting/addresslist-voting.ts:1254]: I assume it means executing only when it has ended and passed?

File: packages/contracts/test/voting/multisig.ts [hunk 11]
```diff
@@ -338,140 +434,189 @@ describe('Multisig', function () {
...
-      );
     });
 
-    it('should set the right onlyListed during initialization', async () => {
-      await multisig.initialize(dao.address, addresslist(5), {
-        onlyListed,
-        minApprovals,
+    describe('execute:', async () => {
+      it('executes if the minimum approval is met', async () => {
+        await multisig.connect(signers[0]).approve(id, false);
```
Comment (brickpop) [inline packages/contracts/test/voting/multisig.ts:519]: Shouldn't there be a check for execute before all 3 approvals, and assert that it reverts then?

File: packages/contracts/test/voting/multisig.ts [hunk 12]
```diff
@@ -338,140 +434,189 @@ describe('Multisig', function () {
...
-          minApprovals,
-        })
-      )
-        .to.emit(multisig, MULTISIG_EVENTS.MULTISIG_SETTINGS_UPDATED)
-        .withArgs(onlyListed, minApprovals);
-    });
+      it('executes if the minimum approval is met and can be called by an unlisted accounts', async () => {
+        await multisig.connect(signers[0]).approve(id, false);
+        await multisig.connect(signers[1]).approve(id, false);
+        await multisig.connect(signers[2]).approve(id, false);
```
Comment (brickpop) [inline packages/contracts/test/voting/multisig.ts:539]: I would also check that execution attempts fail before the actual approvals

File: packages/contracts/test/voting/addresslist-voting.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/addresslist-voting.ts:1254]: Clarified in 49e9ee592bef6d0199ced621ad08a00c93e009c3

File: packages/contracts/test/voting/addresslist-voting.ts [hunk 13]
```diff
@@ -1200,5 +1220,77 @@ describe('AddresslistVoting', function () {
...
+
+        expect(await voting.participation(id)).to.be.gte(
+          votingSettings.minParticipation
+        );
+        expect(await voting.worstCaseSupport(id)).to.be.gt(
+          votingSettings.supportThreshold
+        );
+        expect(await voting.canExecute(id)).to.equal(true);
+
+        // executes normally
```
[the remark this answers is not in the document]
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/addresslist-voting.ts]: Clarified in 49e9ee592bef6d0199ced621ad08a00c93e009c3

**Review by mathewmeconry — DISMISSED** [review]

**Review by novaknole — DISMISSED** [review]

File: packages/contracts/test/voting/multisig.ts (hunk 11 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/multisig.ts:533]: Added a test to check this separately f6ef2ab2e12930b80758c64f0ac107b6c1248a0d

File: packages/contracts/test/voting/multisig.ts (hunk 12 above)
Comment (Michael-A-Heuer) [reply packages/contracts/test/voting/multisig.ts:553]: Added a test to check this separately f6ef2ab2e12930b80758c64f0ac107b6c1248a0d

**Review by novaknole — APPROVED** [review]

**Review by brickpop — APPROVED** [review]
